ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixParticipant.java

```diff
@@ -415,6 +437,27 @@ private boolean removeOldReplicaInfo(ReplicaId replicaId, InstanceConfig instanc
     return removalResult;
   }
 
+  /**
+   * Wait until disabling partition process has completed. This is to avoid race condition where server and Helix may
+   * modify same InstanceConfig.
+   * TODO remove this method after migrating ambry to PropertyStore (in Helix).
+   * @throws InterruptedException
+   */
+  private void awaitDisablingPartition() throws InterruptedException {
+    Properties properties = new Properties();
+    properties.setProperty("helix.property.store.root.path", "/" + clusterName + "/" + PROPERTYSTORE_STR);
+    HelixPropertyStoreConfig propertyStoreConfig = new HelixPropertyStoreConfig(new VerifiableProperties(properties));
+    HelixPropertyStore<ZNRecord> helixPropertyStore =
+        CommonUtils.createHelixPropertyStore(zkConnectStr, propertyStoreConfig, null);
+    String path = PARTITION_DISABLED_ZNODE_PATH + instanceName;
+    int count = 1;
+    while (!helixPropertyStore.exists(path, AccessOption.PERSISTENT)) {
```

Comment:
Do these paths have to be manually removed using a zookeeper client after the decommission is complete and all replicas have called removeOldReplicaInfo()?

What if, instead, `HelixBootStrapUpgradeTool` created a znode before it begins updating using `HelixAdmin` and then deletes it after it is finished? Than, this loop could wait for the znode to stop existing.